bpf/go_sql.c

```diff
@@ -0,0 +1,139 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "headers/bpf_tracing.h"
+#include "common.h"
+#include "bpf_helpers.h"
+#include "bpf_dbg.h"
+#include "utils.h"
+#include <stdbool.h>
+
+#include "arguments.h"
+#include "span_context.h"
+#include "go_context.h"
+//#include "go_types.h"
+//#include "uprobe.h"
+
+char __license[] SEC("license") = "Dual MIT/GPL";
+
+#define MAX_QUERY_SIZE 100
+#define MAX_CONCURRENT 50
+
+#define TRACE_ID_SIZE 16
+#define TRACE_ID_STRING_SIZE 32
+#define SPAN_ID_SIZE 8
+#define SPAN_ID_STRING_SIZE 16
+
+#define BASE_SPAN_PROPERTIES \
+    u64 start_time;          \
+    u64 end_time;            \
+    struct span_context sc;  \
+    struct span_context psc; 
+
+struct sql_request_t {
+    BASE_SPAN_PROPERTIES
+    char query[MAX_QUERY_SIZE];
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__type(key, void*);
+	__type(value, struct sql_request_t);
+	__uint(max_entries, MAX_CONCURRENT);
+} sql_events SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+} events SEC(".maps");
+
+// Injected in init
+volatile const bool should_include_db_statement;
+
+// This instrumentation attaches uprobe to the following function:
+// func (db *DB) queryDC(ctx, txctx context.Context, dc *driverConn, releaseConn func(error), query string, args []any)
+SEC("uprobe/queryDC")
+int uprobe_queryDC(struct pt_regs *ctx) {
```

Comment:
This looks like a simple enough probe, could we not write our own code for fetching a few fields of an object, much like the other Go uprobes we have? I feel like taking this code verbatim as is brings a lot of unnecessary other code with it, like the whole reading of context information etc. None of that is relevant to Beyla. Looks to me like 20-30 lines of code max that need to be written. WDYT?

Reply:
Updated to adapt just the core functionality, and use Beyla ring buffer.